iavl/iavl_path.go

```diff
@@ -0,0 +1,92 @@
+package iavl
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/tendermint/go-wire/data"
+)
+
+// PathToKey represents an inner path to a leaf node.
+// Note that the nodes are ordered such that the last one is closest
+// to the root of the tree.
+type PathToKey struct {
+	LeafHash   data.Bytes           `json:"leaf_hash"`
+	InnerNodes []IAVLProofInnerNode `json:"inner_nodes"`
+}
+
+func (p *PathToKey) String() string {
+	str := ""
+	for i := len(p.InnerNodes) - 1; i >= 0; i-- {
+		str += p.InnerNodes[i].String() + "\n"
+	}
+	str += fmt.Sprintf("hash(leaf)=%s\n", p.LeafHash)
+	return str
+}
+
+func (p *PathToKey) verify(leafNode IAVLProofLeafNode, root []byte) error {
+	leafHash := leafNode.Hash()
+	if !bytes.Equal(leafHash, p.LeafHash) {
+		return InvalidPathErr
+	}
+	hash := leafHash
+	for _, branch := range p.InnerNodes {
+		hash = branch.Hash(hash)
+	}
+	if !bytes.Equal(root, hash) {
+		return InvalidPathErr
+	}
+	return nil
+}
+
+func (p *PathToKey) isLeftmost() bool {
+	for _, node := range p.InnerNodes {
+		if len(node.Left) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (p *PathToKey) isRightmost() bool {
+	for _, node := range p.InnerNodes {
+		if len(node.Right) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (p *PathToKey) isEmpty() bool {
+	return p == nil || len(p.InnerNodes) == 0
+}
+
+func (p *PathToKey) dropRoot() *PathToKey {
+	if p.isEmpty() {
+		return p
+	}
+	return &PathToKey{
+		LeafHash:   p.LeafHash,
+		InnerNodes: p.InnerNodes[:len(p.InnerNodes)-1],
+	}
+}
+
+func (p *PathToKey) hasCommonRoot(p2 *PathToKey) bool {
```

Comment:
@cloudhead this code looks really familiar, like from a review from a week or so ago, is this just a move or reimplementation?